pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,500 @@
+
+	// workerLoopPeriod is the time between worker runs. The workers process the queue of service and ipRange changes.
+	workerLoopPeriod time.Duration
+
+	// tree store the ServiceCIDRs names associated to each
+	muTree sync.Mutex
+	treeV4 *iptree.Tree[cidrSet]
+	treeV6 *iptree.Tree[cidrSet]
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	logger := klog.FromContext(ctx)
+
+	logger.Info("Starting", "controller", controllerName)
+	defer logger.Info("Shutting down", "controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.serviceCIDRsSynced, c.ipAddressSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.worker, c.workerLoopPeriod)
+	}
+	<-ctx.Done()
+}
+
+func (c *Controller) addServiceCIDR(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) updateServiceCIDR(oldObj, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// deleteServiceCIDR
+func (c *Controller) deleteServiceCIDR(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// addIPAddress may block a ServiceCIDR deletion
+func (c *Controller) addIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		return
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// deleteIPAddress may unblock a ServiceCIDR deletion
+func (c *Controller) deleteIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return
+		}
+		ip, ok = tombstone.Obj.(*networkingapiv1alpha1.IPAddress)
+		if !ok {
+			return
+		}
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// ipToCidrs, given an IPAddress return the ServiceCIDRs with the longest prefix match
+// that contains the IP, as it may block or be blocking the deletion of the ServiceCIDR
+func (c *Controller) ipToCidrs(ip *networkingapiv1alpha1.IPAddress) []string {
+	// only process IPs managed by the kube-apiserver
+	managedBy, ok := ip.Labels[networkingapiv1alpha1.LabelManagedBy]
+	if !ok || managedBy != ipallocator.ControllerName {
+		return []string{}
+	}
+
+	ipFamily, ok := ip.Labels[networkingapiv1alpha1.LabelIPAddressFamily]
+	if !ok {
+		return []string{}
+	}
+
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	// the only serviceCIDRs that can be affected by an IPAddress deletiong
+	// are the ones with the longest prefix match.
```

Comment:
Hm... I'm having a hard time convincing myself that this is true for all combinations of overlapping ServiceCIDR Adds and Deletes, but I can't come up with an example where it fails...

Reply:
I put a lot of thought on all the edge cases and I still have that feeling that something is missing, what can I say 🤷

Reply:
ok, this has the issue you commented on the allocator, if CIDR1 has 192.168.0.0/16 and CIDR2 has 192.168.1.0/24, and there is only one IP 192.168.1.0, it should not block the deletion of CIDR2 but block the deletion of CIDR1

Reply:
fixed